source/scoped-mutex/src/lib.rs

```diff
@@ -20,18 +22,32 @@ use core::cell::UnsafeCell;
 ///
 /// Use [`ThreadModeRawMutex`] when data is shared between tasks running on the same executor but you want a global singleton.
 ///
+/// Use [`LockApiRawMutex`] if you wish to use a type implementing
+/// [`lock_api::RawMutex`] as the mutex implementation.
+///
 /// In all cases, the blocking mutex is intended to be short lived and not held across await points.
 ///
 /// [`CriticalSectionRawMutex`]: crate::raw_impls::cs::CriticalSectionRawMutex
 /// [`LocalRawMutex`]: crate::raw_impls::local::LocalRawMutex
-/// [`ThreadModeRawMutex`]: crate::raw_impls::single_core_thread_mode::ThreadModeRawMutex
+/// [`ThreadModeRawMutex`]:
+///     crate::raw_impls::single_core_thread_mode::ThreadModeRawMutex
+/// [`LockApiRawMutex`]: crate::raw_impls::lock_api::LockApiRawMutex
+/// [`lock_api::RawMutex`]:
+///     https://docs.rs/lock_api/0.4.0/lock_api/trait.RawMutex.html
 pub struct BlockingMutex<R, T: ?Sized> {
     // NOTE: `raw` must be FIRST, so when using ThreadModeMutex the "can't drop in non-thread-mode" gets
     // to run BEFORE dropping `data`.
     raw: R,
     data: UnsafeCell<T>,
 }
 
+/// A RAII guard that allows access to the data guarded by a [`BlockingMutex`].
+#[must_use]
+pub struct MutexGuard<'mutex, R: RawMutex, T: ?Sized> {
+    data: &'mutex mut T,
```

Comment:
Should this be `NonNull<T>` instead, because if we were to unlock the mutex in the drop impl, we could technically momentarily have aliased mutable references?

For some reason I've been in the habit of always using pointers inside of guard-like structures, and I think this is the reason, but I can't cite a specific ruling for this.

Reply:
hmm, yeah, that might be more correct!

my initial implementation actually just had a `&'mutex BlockingMutex<R, T>` and used `UnsafeCell::get` every time it dereferenced the guard, but i did this instead because i thought it was maybe nice to avoid offsetting a pointer on every deref --- which is almost certainly a premature optimization! `NonNull` would solve that problem while also avoiding concurrent mutable refs...

Reply:
Yeah, I guess that's a size/speed tradeoff, tho maybe the optimizer can see through the basic pointer offset so it isn't necessary to keep two pointers that are always right next to each other? Dunno, but it's also probably something we can do in a non-breaking change.

Reply:
**:musical_note:  Let's See What `std` Does! :notes:**

Turns out it actually does the "guard just owns a `&Mutex` thing:  

```rust
pub struct MutexGuard<'a, T: ?Sized + 'a> {
    lock: &'a Mutex<T>,
    poison: poison::Guard,
}
```

(see https://doc.rust-lang.org/stable/src/std/sync/mutex.rs.html#209-212)

That actually _might_ be the best, because it makes the `MutexGuard` a single word instead of two words. On x86, at least, they're both a single `mov` instruction, and I don't *think* the address-with-offset `mov` is more expensive than the mov without offset: https://godbolt.org/z/jnz19errj

Reply:
Yeah, on thumb I think you can do a load with an IMM offset in a register, so it's probably whatever there too.

Reply:
...actually, upon going back and making the godbolt a bit more representative, it looks like the "pointer to data" guard generates a single-`mov` deref-like function, but the "pointer to lock" guard generates a `mov` and an `add`, which feels unfortunate: https://godbolt.org/z/719GvfKKM

so, i guess we have a tradeoff between a two-word `MutexGuard` or a two-instruction `Deref`